fv/workload/workload.go

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2017 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package workload
+
+import (
+	"fmt"
+	"os/exec"
+	"strings"
+
+	. "github.com/onsi/gomega"
+	"github.com/projectcalico/felix/fv/containers"
+	"github.com/projectcalico/libcalico-go/lib/api"
+	"github.com/projectcalico/libcalico-go/lib/client"
+	"github.com/projectcalico/libcalico-go/lib/net"
+	log "github.com/sirupsen/logrus"
+)
+
+type Workload struct {
+	C             *containers.Container
+	Name          string
+	InterfaceName string
+	IP            string
+	Port          string
+	Stop          func()
+}
+
+var workloadIdx = 0
+
+func Run(c *containers.Container, interfaceName, ip, port string) (w *Workload) {
+
+	// Build unique workload name and struct.
+	workloadIdx++
+	w = &Workload{
+		C:             c,
+		Name:          fmt.Sprintf("w%v", workloadIdx),
+		InterfaceName: interfaceName,
+		IP:            ip,
+		Port:          port,
+	}
+
+	// Start the workload.
+	log.WithField("workload", w).Info("About to run workload")
+	exec.Command("docker", "cp", "../bin/test-workload",
+		w.C.Name+":/test-workload").Run()
+	workloadCmd := exec.Command("docker", "exec", w.C.Name,
+		"sh", "-c",
+		fmt.Sprintf("echo $$ > /tmp/%v; exec /test-workload %v %v %v",
+			w.Name,
+			w.InterfaceName,
+			w.IP,
+			w.Port))
+	err := workloadCmd.Start()
+	Expect(err).NotTo(HaveOccurred())
+
+	// Fill in rest of container struct.
+	w.Stop = func() {
+		outputBytes, err := exec.Command("docker", "exec", w.C.Name,
+			"cat",
+			fmt.Sprintf("/tmp/%v", w.Name)).CombinedOutput()
+		Expect(err).NotTo(HaveOccurred())
```

Comment:
If you hit this error, you won't shut the container down properly

Reply:
I'm afraid I don't see the problem.  Note that the code here is about killing an additional process (representing a test workload) that is running inside a container; not about killing the container itself.

Could you explain further?

Reply:
Sorry, was confusing this with Container.Stop.  Does it matter that you won't do the kill if this Expect fails?

Reply:
No, I don't think so, because AfterSuite will always run (regardless of any previous failures) and will rm -f all the containers, and that will also clean up any additional processes inside those containers.